pwiz_tools/SeeMS/Dialogs/HeatmapForm.cs

```diff
@@ -250,22 +250,53 @@ private void ShowHeatmap(double scanTime, int msLevel)
             heatmapGraphPane.CurveList.Clear();
             heatmapGraphPane.GraphObjList.Add(new TextObj("Loading...", 0.5, 0.5, CoordType.ChartFraction) {FontSpec = new FontSpec {Border = new Border {IsVisible = false}, IsBold = true, Size = 24}});
             msGraphControl.Refresh();
-
+            
+            var mainSpectrumList = source.Source.MSDataFile.run.spectrumList;
             var bounds = heatmapBoundsByMsLevel[msLevel];
+            int lastBinIndex = -1;
             foreach (var bin in ionMobilityBins)
             {
-                var spectrum = source.GetMassSpectrum(bin.BinIndex);
-                var points = spectrum.Points;
-                for (int j = 0; j < points.Count; ++j)
+                // skip bins where the spectrum is the same as the last bin (that spectrum's points were added from its mobility array)
+                if (bin.BinIndex == lastBinIndex)
+                    continue;
+                lastBinIndex = bin.BinIndex;
+
+                var s = mainSpectrumList.spectrum(bin.BinIndex, true);
+                var mzArray = s.getMZArray().data;
+                var intensityArray = s.getIntensityArray().data;
+                var mobilityBDA = s.getArrayByCVID(pwiz.CLI.cv.CVID.MS_mean_drift_time_array);
+                if (mobilityBDA == null)
+                    mobilityBDA = s.getArrayByCVID(pwiz.CLI.cv.CVID.MS_mean_ion_mobility_array);
+
+                if (mobilityBDA != null)
                 {
-                    double mz = points[j].X;
-                    double intensity = points[j].Y;
-                    bounds.MinX = Math.Min(bounds.MinX, mz);
-                    bounds.MinY = Math.Min(bounds.MinY, bin.IonMobility);
-                    bounds.MaxX = Math.Max(bounds.MaxX, mz);
-                    bounds.MaxY = Math.Max(bounds.MaxY, bin.IonMobility);
-
-                    heatmapPoints.Add(new Point3D(mz, bin.IonMobility, intensity));
+                    var mobilityArray = mobilityBDA.data;

```

Comment:
@bspratt Here's where I am adding each mz/mobility/intensity point to the heatmap. The heatmap has to know all the mobility bins in both cases, but with the combined representation, it populates them all from a single spectrum (for MS1; for PASEF MS2, there will still be multiple spectra, but far fewer).